File: matlab/utilities/input/set_default_options.m
function opt = set_default_options(opt, nunits)
% SET_DEFAULT_OPTIONS  Fill in any missing options with defaults
%
%   opt = set_default_options(opt, nunits) takes a partial options struct
%   and fills in any missing fields with default values. Also normalizes
%   string inputs to char arrays for consistent strcmp behavior.
%
% -------------------------------------------------------------------------
% Inputs:
% -------------------------------------------------------------------------
%
% <opt> - partial struct with user-specified options. May be missing fields.
%
% <nunits> - number of units in the data (used for default unit_groups)
%
% -------------------------------------------------------------------------
% Returns:
% -------------------------------------------------------------------------
%
% <opt> - complete struct with all required fields. Defaults are:
%   basis              = 'signal'
%   criterion          = 'prediction'
%   threshold_method   = 'hybrid'
%   basis_ordering     = 'eigenvalues'
%   variance_threshold = 0.99
%   allowable_thresholds = []
%   unit_groups        = (1:nunits)' for hybrid/unit modes, zeros for global
%   gsn_args           = struct()
%   wantfig            = 1
%   figurepath         = '' (empty = display only, no save)
%   wantverbose        = 1

    % Normalize string inputs to char for consistent strcmp behavior
    % MATLAB string type ("text") vs char array ('text') behave differently with ischar()
    if isfield(opt, 'basis') && isstring(opt.basis)
        opt.basis = char(opt.basis);
    end
    if isfield(opt, 'criterion') && isstring(opt.criterion)
        opt.criterion = char(opt.criterion);
    end
    if isfield(opt, 'threshold_method') && isstring(opt.threshold_method)
        opt.threshold_method = char(opt.threshold_method);
    end
    if isfield(opt, 'basis_ordering') && isstring(opt.basis_ordering)
        opt.basis_ordering = char(opt.basis_ordering);
    end

    if ~isfield(opt, 'basis')
        opt.basis = 'signal';
    end

    if ~isfield(opt, 'criterion')
        opt.criterion = 'prediction';
    end

    if ~isfield(opt, 'threshold_method')
        opt.threshold_method = 'hybrid';
    end

    if ~isfield(opt, 'basis_ordering')
        opt.basis_ordering = 'eigenvalues';
    end

    if ~isfield(opt, 'variance_threshold')
        opt.variance_threshold = 0.99;
    end

    if ~isfield(opt, 'allowable_thresholds')
        opt.allowable_thresholds = [];
    end

    if ~isfield(opt, 'unit_groups')
        if strcmp(opt.threshold_method, 'global')
            opt.unit_groups = zeros(nunits, 1);
        else
            opt.unit_groups = (1:nunits)';
        end
    end

    if ~isfield(opt, 'gsn_args') || isempty(opt.gsn_args) || ~isstruct(opt.gsn_args)
        opt.gsn_args = struct();
    end

    if ~isfield(opt, 'wantfig')
        opt.wantfig = 1;
    end

    if ~isfield(opt, 'wantverbose')
        opt.wantverbose = 1;
    end

    if ~isfield(opt, 'figurepath')
        opt.figurepath = '';
    end

    % Auto-detect: if allowable_thresholds is a single value, force threshold_method to 'global'
    if ~isempty(opt.allowable_thresholds)
        if isscalar(opt.allowable_thresholds)
            if ~strcmp(opt.threshold_method, 'global')
                if opt.wantverbose
                    fprintf('PSN: allowable_thresholds is a single value, automatically setting threshold_method to ''global''\n');
                end
                opt.threshold_method = 'global';
                % Update unit_groups to match global mode
                opt.unit_groups = zeros(nunits, 1);
            end
        end
    end

    validate_options(opt, nunits);
end

File: psn/utilities/input/set_default_options.py
"""Options validation and defaults for PSN."""

import numpy as np

from .validate_options import validate_options as _validate_options


def set_default_options(opt, nunits):
    """SET_DEFAULT_OPTIONS  Fill in any missing options with defaults

    opt = set_default_options(opt, nunits) takes a partial options dict
    and fills in any missing fields with default values.

    -------------------------------------------------------------------------
    Inputs:
    -------------------------------------------------------------------------

    <opt> - partial dict with user-specified options. May be missing fields.

    <nunits> - number of units in the data (used for default unit_groups)

    -------------------------------------------------------------------------
    Returns:
    -------------------------------------------------------------------------

    <opt> - complete dict with all required fields. Defaults are:
      basis              = 'signal'
      criterion          = 'max-tradeoff'
      threshold_method   = 'global'
      basis_ordering     = 'eigenvalues'
      variance_threshold = 0.99
      allowable_thresholds = None
      unit_groups        = np.arange(nunits) for hybrid mode, zeros for global
      alpha              = None (disabled; interpolates between prediction and variance)
      gsn_result         = None (pass previous results['gsn_result'] to skip GSN)
      gsn_args           = {}
      wantfig            = True
      wantverbose        = True
      figurepath         = None
      cmap               = None (uses cmapsign4 in visualization)
      skip_split_half    = False (keep analytic recovery curve but skip the
                           costly empirical split-half tradeoff computation)
    """

    # Create a copy to avoid modifying the original
    opt = opt.copy()

    if 'basis' not in opt:
        opt['basis'] = 'signal'

    if 'criterion' not in opt:
        opt['criterion'] = 'max-tradeoff'

    if 'threshold_method' not in opt:
        opt['threshold_method'] = 'global'

    if 'basis_ordering' not in opt:
        opt['basis_ordering'] = 'eigenvalues'

    if 'variance_threshold' not in opt:
        opt['variance_threshold'] = 0.99

    if 'allowable_thresholds' not in opt:
        opt['allowable_thresholds'] = None

    if 'unit_groups' not in opt:
        if opt['threshold_method'] == 'global':
            opt['unit_groups'] = np.zeros(nunits, dtype=int)
        else:
            opt['unit_groups'] = np.arange(nunits)

    if 'gsn_result' not in opt:
        opt['gsn_result'] = None

    if 'gsn_args' not in opt or opt['gsn_args'] is None or not isinstance(opt['gsn_args'], dict):
        opt['gsn_args'] = {}

    if 'wantfig' not in opt:
        opt['wantfig'] = True

    if 'wantverbose' not in opt:
        opt['wantverbose'] = True

    if 'figurepath' not in opt:
        opt['figurepath'] = None

    if 'cmap' not in opt:
        opt['cmap'] = None  # Will use default cmapsign4 in visualization

    if 'split_half_metric' not in opt:
        opt['split_half_metric'] = 'correlation'  # 'correlation' or 'mse'

    if 'alpha' not in opt:
        opt['alpha'] = None

    if 'skip_split_half' not in opt:
        opt['skip_split_half'] = False

    # Auto-detect: if allowable_thresholds is a single value, force threshold_method to 'global'
    if opt['allowable_thresholds'] is not None:
        allowable_arr = np.asarray(opt['allowable_thresholds'])
        if allowable_arr.size == 1:            # bare scalar or length-1 vector
            if opt['threshold_method'] != 'global':
                if opt['wantverbose']:
                    print("PSN: allowable_thresholds is a single value, automatically setting threshold_method to 'global'")
                opt['threshold_method'] = 'global'
                # Update unit_groups to match global mode
                opt['unit_groups'] = np.zeros(nunits, dtype=int)

    # Note: conflicts between a Wiener request and the basis/criterion/threshold
    # pipeline are rejected up front in parse_inputs (see _reject_wiener_conflicts),
    # where the user-supplied keys are still distinguishable from filled defaults.

    _validate_options(opt, nunits)
    return opt
